### PR TITLE
Add possibility to archive project

### LINKED ISSUE #31: Add possibility to archive project
Archived projects shouldn't be present in the left menu bar. At the bottom of the project-selection menu  should be an icon to display/collapse also archived projects.

Project should be archived by writing msg `(archive :__cledara)` to the #pm-reports channel. The command should be able to take more projects: e.g. `(archive :__prj1 :__prj2)` . Implement also the `unarchive` function with the same signature.

Make sure that the parsing mechanism can handle whitespaces (including newlines) in between the words.

As for the UI, I don't mind if the same effect ((un)archiving projects) is achieved via standard fronted, but I think that parsing messages from Slack is something we already do, so why not use it.

### PR DESCRIPTION
**Pull request by leanil**
Store archived status for each tag in a new table.
Archived projects are not visible on UI, unless searching for them or checkbox is ticked.
    
Resolves #31.

### CODE REVIEW

File: bot/src/slack.js [hunk 1]
```diff
@@ -45,6 +59,8 @@ const updateMessage = async (event) => {
   await upsertReport({ts, message})
   await clearTags(ts)
 
+  if (await handleArchiveCommand(event.message, event.ts)) return
```
Comment (leanil) [inline bot/src/slack.js]: The ts in `event.message` is when the message was originally sent.
I need to update the archived status with the ts of the modification.

File: client/src/components/leftPanel/LeftPanel.js [hunk 2]
```diff
@@ -22,15 +25,12 @@ class LeftPanel extends Component {
...
         </a>
         <strong className="sectionTitle">Select tag</strong>
-        <input
-          type="text"
-          value={searchString}
-          onChange={this.handleChange}
-          placeholder="search"
-        />
+        <QuickSearch value={searchString} onChange={this.onSearchChange} />
+        <ShowArchivedBox label="show archived projects" isChecked={showArchived} onChange={this.onShowArchivedChange} />
```
Comment (leanil) [inline client/src/components/leftPanel/LeftPanel.js]: According to the issue, this box should be at the bottom.
But I didn't find a way to prevent the tags from going under the box,
and putting it on the top feels more consistent, as the controls are in one place.
bottom
image

File: migrations/20200205161120_create_archived_table.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        .boolean('is_archived')
+        .comment('Indicates whether the project is hidden by default.')
+    })
+  }
+  
+  await knex.raw(`
+      INSERT INTO archived(tag,is_archived)
+      SELECT DISTINCT tag, FALSE
+      FROM tag;
+  `)
```
Comment (paldys) [inline migrations/20200205161120_create_archived_table.js]: Is there a reason you went with a completely new table instead of using `tag` table? There is a one-to-one relationship between them, isn't there?

File: client/src/components/leftPanel/LeftPanel.js [hunk 4]
```diff
@@ -6,14 +6,17 @@ import './LeftPanel.scss'
 class LeftPanel extends Component {
 
   state = {
-    searchString: ""
+    searchString: "",
+    showArchived: false,
   }
 
-  handleChange = (event) => this.setState({ searchString: event.target.value })
+  onSearchChange = (event) => this.setState({ searchString: event.target.value })
```
Comment (paldys) [inline client/src/components/leftPanel/LeftPanel.js:13]: Please, move refactors into a separate commit.

File: client/src/components/leftPanel/LeftPanel.js (hunk 2 above)
Comment (paldys) [reply client/src/components/leftPanel/LeftPanel.js]: >  But I didn't find a way to prevent the tags from going under the box,

There is always a way. :slightly_smiling_face: 

What you are saying is not wrong (controls in one place), but I don't think it looks nice and from UX perspective what and how Tomas suggested could work better. Where your solution fails is the search. Search should work with archived projects too always, so it would not be consistent with what the user sees with the checkbox.

He wrote to have an icon/text on the end which would collapse/display archived messages. This could be hidden when you are searching. Also, I believe some visual extinction (an icon? ) for the archived projects would be also nice.

**Review by paldys — CHANGES_REQUESTED** [review]

File: migrations/20200205161120_create_archived_table.js (hunk 3 above)
Comment (leanil) [reply migrations/20200205161120_create_archived_table.js]: In the `tag` table, the tag is not unique, as each tag can be present on many reports.

Comment (paldys) [reply migrations/20200205161120_create_archived_table.js]: Oh.. I see.

File: server/src/index.js [hunk 5]
```diff
@@ -95,7 +95,7 @@ app.get('/api/reports-by-tags/:tag', authorize, async (req, res) => {
 
   const authors = reports.map((r) => r.user)
   const replyAuthors = [].concat(...reports.map((t) => t.replies.map((r) => r.user)))
-  const allMentions = 
+  const allMentions =
```
Comment (leanil) [inline server/src/index.js]: This will disappear when I rebase.

File: client/src/components/leftPanel/LeftPanel.js [hunk 6]
```diff
@@ -6,37 +6,58 @@ import './LeftPanel.scss'
...
+          </a>
+          <strong className="sectionTitle">Select tag</strong>
+          <input
+            type="text"
+            value={searchString}
+            onChange={this.handleChange}
+            placeholder="search"
+          />
+          {!searchString && <ShowArchivedBox label="show archived projects" isChecked={showArchived}
+                                             onChange={this.onShowArchivedChange} />}
```
Comment (paldys) [inline client/src/components/leftPanel/LeftPanel.js]: (nitpick) Is eslint forcing you to do this formatting? If not I would prefer the following:
```
{!searchString && (
	<ShowArchivedBox
		label="show archived projects"
		isChecked={showArchived}
		onChange={this.onShowArchivedChange}
	/>
)}
```

File: client/src/components/leftPanel/LeftPanel.js [hunk 7]
```diff
@@ -6,37 +6,58 @@ import './LeftPanel.scss'
...
+          <ul>
+            {tags
+              .filter(({tag, isArchived}) => tag.toLowerCase().includes(lowerCaseSearchString) &&
+                                             (searchString || !isArchived || showArchived))
```
Comment (paldys) [inline client/src/components/leftPanel/LeftPanel.js]: (nitpick) again if not eslint is forcing you to do this, let's go with:
```
  tags
    .filter(
      ({tag, isArchived}) =>
        tag.toLowerCase().includes(lowerCaseSearchString) &&
        (searchString || !isArchived || showArchived),
    )
    .map((tag) => <Tag key={tag.tag} tag={tag} />)
```

File: migrations/20200212110207_create_tag_table.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+exports.up = async (knex) => {
+  if (!await knex.schema.hasTable('tag')) {
+    await knex.schema.createTable('tag', (table) => {
+      table
+        .string('tag')
+        .primary()
+        .comment('Emoji identifying some project, like :__bee:.')
+      table
+        .boolean('is_archived')
+        .comment('Indicates whether the project is hidden by default.')
```
Comment (paldys) [inline migrations/20200212110207_create_tag_table.js]: Let's stick with:
```suggestion
        .comment('Indicates whether the project is archived.')
```

**Review by paldys — COMMENTED** [review]
Looks good, just a few nitpicks.

Also, please fix _Add possibility to archive project_ commit's message, there are few grammar errors (e.g sentence _Store archived..._) and you address some things there which are fixed only in later commits (e.g. _give feedback on recognized commands_).

**Comment by leanil** [comment]
Updated the icon and aligned it vertically.

**Comment by leanil** [comment]
Display commands as regular reports.

**Review by paldys — APPROVED** [review]
:+1:
